parquet-column/src/test/java/org/apache/parquet/column/values/bloom/Bloom.java

```diff
@@ -0,0 +1,413 @@
+      mask[i] = mask[i] >>> 27;
+    }
+
+    for (int i = 0; i < 8; ++i) {
+      mask[i] = 0x1 << mask[i];
+    }
+
+    return mask;
+  }
+
+  /**
+   * Add an element to bloom filter, the element content is represented by
+   * the hash value of its plain encoding result.
+   * @param hash hash result of element.
+   */
+  private void addElement(long hash) {
+    int bucketIndex = (int)(hash >> 32) & (bitset.length / BYTES_PER_BUCKET - 1);
+    int key = (int)hash;
+
+    // Calculate mask for bucket.
+    int mask[] = setMask(key);
+
+    for (int i = 0; i < 8; i++) {
+      int value = intBuffer.get(bucketIndex * (BYTES_PER_BUCKET / 4) + i);
+      value |= mask[i];
+      intBuffer.put(bucketIndex * (BYTES_PER_BUCKET / 4) + i, value);
+    }
+  }
+
+  /**
+   * Determine where an element is in set or not.
+   * @param hash the hash value of element plain encoding result.
+   * @return false if element is must not in set, true if element probably in set.
+   */
+  private boolean isContain(long hash) {
+    int bucketIndex = (int)(hash >> 32) & (bitset.length / BYTES_PER_BUCKET - 1);
+    int key = (int)hash;
+
+    // Calculate mask for bucket.
+    int mask[] = setMask(key);
+
+    for (int i = 0; i < 8; i++) {
+      if (0 == (intBuffer.get(bucketIndex * (BYTES_PER_BUCKET / 4) + i) & mask[i])) {
+        return false;
+      }
+    }
+
+    return true;
+  }
+
+  /**
+   * Calculate optimal size according to the number of distinct values and false positive probability.
+   * @param n: The number of distinct values.
+   * @param p: The false positive probability.
+   * @return optimal number of bits of given n and p.
+   */
+  public static int optimalNumOfBits(long n, double p) {
+    Preconditions.checkArgument((p > 0.0 && p < 1.0),
+      "DEFAULT_FPP should be less than 1.0 and great than 0.0");
+
+    final double M = -8 * n / Math.log(1 - Math.pow(p, 1.0 / 8));
+    final double MAX = ParquetProperties.DEFAULT_MAXIMUM_BLOOM_FILTER_BYTES << 3;
+    int numBits = (int)M;
+
+    // Handle overflow.
+    if (M > MAX || M < 0) {
+      numBits = (int)MAX;
+    }
+
+    // Get next power of 2 if bits is not power of 2.
+    if ((numBits & (numBits - 1)) != 0) {
+      numBits = Integer.highestOneBit(numBits) << 1;
+    }
+
+    // Minimum
+    if (numBits < (BYTES_PER_BUCKET << 3)) {
+      numBits = BYTES_PER_BUCKET << 3;
+    }
+
+    return numBits;
+  }
+
+  /**
+   * used to decide if we want to work to the next page
+   * @return Bytes buffered of bloom filter.
+   */
+  public long getBufferedSize() {
+    return bitset.length;
+  }
+
+  /**
+   * Compute hash for int value by using its plain encoding result.
+   * @param value the value to hash
+   * @return hash result
+   */
+  public long hash(int value) {
+    ByteBuffer plain = ByteBuffer.allocate(Integer.SIZE/Byte.SIZE);
+    plain.order(ByteOrder.LITTLE_ENDIAN).putInt(value);
+    switch (hash) {
+      case MURMUR3_X64_128: return Murmur3.hash64(plain.array());
```

Comment:
This should hash a `ByteBuffer` instead of `byte[]`. That will require less copying for types like `Binary`.

Reply:
guava support ByteBuffer argument from version 23. I just update pom.xml also .

Reply:
Looks like guava 23.0 changes quite some APIs which cause several build issues in parquet-cli.  I reverted back to use guava 20.0 to use hash64(byte[]) API firstly.